fuse_core/src/async_motion_model.cpp

```diff
@@ -62,56 +62,79 @@ bool AsyncMotionModel::apply(Transaction& transaction)
+  auto callback = std::make_shared<CallbackWrapper<bool>>(
     std::bind(&AsyncMotionModel::applyCallback, this, std::ref(transaction)));
   auto result = callback->getFuture();
-  callback_queue_.addCallback(callback, reinterpret_cast<uint64_t>(this));
+  callback_queue_->addCallback(callback);
   result.wait();
-  return result.get();
-}
 
-void AsyncMotionModel::graphCallback(Graph::ConstSharedPtr graph)
-{
-  callback_queue_.addCallback(
-    boost::make_shared<CallbackWrapper<void>>(std::bind(&AsyncMotionModel::onGraphUpdate, this, std::move(graph))),
-    reinterpret_cast<uint64_t>(this));
+  return result.get();
 }
 
 void AsyncMotionModel::initialize(const std::string& name)
 {
   // Initialize internal state
   name_ = name;
-  node_handle_.setCallbackQueue(&callback_queue_);
-  private_node_handle_ = ros::NodeHandle("~/" + name_);
-  private_node_handle_.setCallbackQueue(&callback_queue_);
+  std::string node_namespace = "";
+
+  // TODO(CH3): Pass in the context or a node to get the context from
+  rclcpp::Context::SharedPtr ros_context = rclcpp::contexts::get_global_default_context();
+  auto node_options = rclcpp::NodeOptions();
+  node_options.context(ros_context); //set a context to generate the node in
+
+  // TODO(CH3): Potentially pass in the optimizer node instead of spinning a new one
+  node_ = rclcpp::Node::make_shared(name_, node_namespace, node_options);
+
+  auto executor_options = rclcpp::ExecutorOptions();
+  executor_options.context = ros_context;
+  executor_ = rclcpp::executors::MultiThreadedExecutor::make_shared(executor_options, executor_thread_count_);
+
+  callback_queue_ = std::make_shared<CallbackAdapter>(ros_context);
+  node_->get_node_waitables_interface()->add_waitable(
+    callback_queue_, (rclcpp::CallbackGroup::SharedPtr) nullptr);
 
   // Call the derived onInit() function to perform implementation-specific initialization
   onInit();
 
-  // Start the async spinner to service the local callback queue
-  spinner_.start();
+  executor_->add_node(node_);
+}
+
+void AsyncMotionModel::graphCallback(Graph::ConstSharedPtr graph)
+{
+  auto callback = std::make_shared<CallbackWrapper<void>>(
+    std::bind(&AsyncMotionModel::onGraphUpdate, this, std::move(graph))
+  );
+  // auto result = callback->getFuture();  // TODO(CH3): Circle back if we need this again
+  callback_queue_->addCallback(callback);
+  // result.wait();
 }
 
 void AsyncMotionModel::start()
 {
-  auto callback = boost::make_shared<CallbackWrapper<void>>(std::bind(&AsyncMotionModel::onStart, this));
-  auto result = callback->getFuture();
-  callback_queue_.addCallback(callback, reinterpret_cast<uint64_t>(this));
-  result.wait();
+  auto callback = std::make_shared<CallbackWrapper<void>>(
+    std::bind(&AsyncMotionModel::onStart, this)
+  );
+  // auto result = callback->getFuture();  // TODO(CH3): Circle back if we need this again
+  callback_queue_->addCallback(callback);
+  // result.wait();
```

Comment:
I think it's needed in `start` - at least that's what the ROS 1 code is doing. It looks like starting/stopping is synchronous and graphCallback is asynchronous in the ROS 1 code.

Reply:
https://github.com/locusrobotics/fuse/compare/d3b6c944272008f0ad4f62582c73deb350f3b9e1..3e38f8d4c103da15b673464842e504c4d7a8f799